build.assets/tooling/cmd/resource-ref-generator/reference/reference.go

```diff
@@ -0,0 +1,332 @@
+			// If the type of the field is an *ast.SelectorExpr,
+			// it's of the form <package>.<type name>.
+			g, ok := t.X.(*ast.Ident)
+			if ok {
+				gopkg = g.Name
+			}
+			fldname = t.Sel.Name
+
+		// There's no package, so only assign a name.
+		case *ast.Ident:
+			fldname = t.Name
+		}
+
+		for _, ti := range finalTypes {
+			if gopkg == ti.Package && fldname == ti.Name {
+				m = true
+				break
+			}
+		}
+	}
+
+	return m
+}
+
+type GenerationError struct {
+	messages []error
+}
+
+func (g GenerationError) Error() string {
+	// Begin with a newline to format the first list item below the outer
+	// error.
+	final := "\n"
+	for _, e := range g.messages {
+		final += fmt.Sprintf("- %v\n", e)
+	}
+	return final
+}
+
+// Generate uses the provided user-facing configuration to write the resource
+// reference to fs.
+func Generate(srcFS, destFS afero.Fs, conf GeneratorConfig) error {
+	sourceData, err := resource.NewSourceData(srcFS, conf.SourcePath)
+	if err != nil {
+		return fmt.Errorf("can't load Go source files: %v", err)
+	}
+
+	versionKindAssignments, err := resource.VersionKindAssignments(sourceData.PossibleFuncDecls, conf.FieldAssignmentMethodName)
+	if err != nil {
+		return err
+	}
+
+	// Extract data from a declaration to transform it into a reference
+	// entry later
+	errs := GenerationError{messages: []error{}}
+	for k, decl := range sourceData.TypeDecls {
+		if !shouldProcess(decl, conf.RequiredFieldTypes, conf.ExcludedResourceTypes) {
+			continue
+		}
+
+		pc := pageContent{}
+
+		// decl is a dynamic resource type, so get data for the type and
+		// its dependencies.
+		entries, err := resource.ReferenceDataFromDeclaration(decl, sourceData.TypeDecls)
+		if errors.Is(err, resource.NotAGenDeclError{}) {
+			continue
+		}
+		if err != nil {
+			errs.messages = append(errs.messages, fmt.Errorf("issue creating a reference entry for declaration %v.%v in file %v: %v", k.PackageName, k.DeclName, decl.FilePath, err))
+		}
+
+		pc.Resource.ReferenceEntry = entries[k]
+		delete(entries, k)
+		pc.Fields = entries
+
+		vk, ok := versionKindAssignments[k]
+		var verName, kindName string
+		if ok {
+			// So far, all values of "Kind" and "Version"
+			// are declared in the same package as the types
+			// that include these fields.
+			verName = sourceData.StringAssignments[resource.PackageInfo{
+				DeclName:    vk.Version,
+				PackageName: k.PackageName,
+			}]
+
+			kindName = sourceData.StringAssignments[resource.PackageInfo{
+				DeclName:    vk.Kind,
+				PackageName: k.PackageName,
+			}]
+		}
+		pc.Resource.Kind = kindName
+		pc.Resource.Version = verName
+
+		filename := strings.ReplaceAll(strings.ToLower(pc.Resource.SectionName), " ", "-")
+		doc, err := destFS.Create(path.Join(conf.DestinationDirectory, filename+".mdx"))
+		if err != nil {
+			errs.messages = append(errs.messages, err)
+		}
+
+		if err := template.Must(template.New("Main reference").Parse(referenceTemplate)).Execute(doc, pc); err != nil {
```

Comment:
I would parse the template once outside of the loop (potentially even in an init func) and then reuse it here. Otherwise we're wasting time re-parsing the same template multiple times.